# What are Recommender Systems?

Companies like Amazon, Netflix, Linkedin, and Pandora leverage recommender systems to help users discover new and relevant items (products, videos, jobs, music), creating a delightful user experience while driving incremental revenue.

# User Based Collaborative Filtering

These are recommendations based on past behavior.

Collaborative filtering, by the way, is just a fancy name for saying recommending stuff based on the combination of what you did and what everybody else did, okay? So, it's looking at your behavior and comparing that to everyone else's behavior, to arrive at the things that might be interesting to you that you haven't heard of yet.

1. The idea here is we build up a matrix of everything that every user has ever bought, or viewed, or rated, or whatever signal of interest that you want to base the system on. So basically, we end up with a row for every user in our system, and that row contains all the things they did that might indicate some sort of interest in a given product. So, picture a table, I have users for the rows, and each column is an item, okay? That might be a movie, a product, a web page, whatever; you can use his for many different things.
2. I then use that matrix to compute the similarity between different users. So, I basically treat each row f this as a vector and I can compute the similarity between each vector of users, based on their behavior.
3. Two users who liked mostly the same things would be very similar to each other and I can then sort this by those similarity scores. If I can find all the users similar to you based on their past behavior, I can then find the users most similar to me, and recommend stuff that they liked that I didn't look at yet.

Let's look at a example, and it'll make a little bit more sense.

![title](User_Based_Collaborative_Filtering.PNG)

Let's say that this nice lady in the preceding image watched Star Wars and The Empire Strikes Back and she loved them both. So, we have a user vector, of this lady, giving a 5-star rating to Star Wars and The Empire Strikes Back.

Let's also say Mr. Edgy Mohawk Man comes along and he only watched Star Wars. That's the only thing he's seen, he doesn't know about The Empire Strikes Back yet, somehow, he lives in some strange universe where he doesn't know that there are actually many, many Star Wars movies, growing every year in fact.

We can of course say that this guy's actually similar to this other lady because they both enjoyed Star Wars a lot, so their similarity score is probably fairly good and we can say, okay, well, what has this lady enjoyed that he hasn't seen yet? And, The Empire Strikes Back is one, so we can then take that information that these two users are similar based on their enjoyment of Star Wars, find that this lady also liked The Empire Strikes Back, and then present that as a good recommendation for Mr. Edgy Mohawk Man. We can then go ahead and recommend The Empire Strikes Back to him and he'll probably love it.

## Limitations of User Based Filtering

Now, unfortunately, user-based collaborative filtering has some limitations. When we think about relationships and recommending things based on relationships between items and people and whatnot, our mind tends to go on relationships between people. So, we want to find people that are similar to you and recommend stuff that they liked. That's kind of the intuitive thing to do, but it's not the best thing to do! 

The following is the list of some limitations of user-based collaborative filtering:

1. One problem is that people are fickle; their tastes are always changing. So, maybe that nice lady in the previous example had sort of a brief science fiction action film phase that she went through and then she got over it, and maybe later in her life she started getting more into dramas or romance films or romcoms. So, what would happen if my Edgy Mohawk guy ended up with a high similarity to her just based on her earlier sci-fi period, and we ended up recommending romantic comedies to him as a result? That would be bad. I mean, there is some protection against that in terms of how we compute the similarity scores to begin with, but it still pollutes our data that people's tastes can change over time. So, comparing people to people isn't always a straightforward thing to do, because people change.

2. The other problem is that there's usually a lot more people than there are things in your system, so 7 billion people in the world and counting, there's probably not 7 billion movies in the world, or 7 billion items that you might be recommending out of your catalog. The computational problem finding all the similarities between all of the users in your system is probably much greater than the problem of finding similarities between the items in your system. So, by focusing the system on users, you're making your computational problem a lot harder than it might need to be, because you have a lot of users, at least hopefully you do if you're working for a successful company. The final problem is that people do bad things. There's a very real economic incentive to make sure that your product or your movie or whatever it is gets recommended to people, and there are people who try to game the system to make that happen for their new movie, or their new product, or their new book, or whatever.

3. It's pretty easy to fabricate fake personas in the system by creating a new user and having them do a sequence of events that likes a lot of popular items and then likes your item too. This is called a shilling attack, and we want to ideally have a system that can deal with that.

4. There is research around how to detect and avoid these shilling attacks in user-based collaborative filtering, but an even better approach would be to use a totally different approach entirely that's not so susceptible to gaming the system.

5. That's user-based collaborative filtering. Again, it's a simple concept-you look at similarities between users based on their behavior, and recommend stuff that a user enjoyed that was similar to you, that you haven't seen yet. Now, that does have its limitations as we talked about.

## Item Based Collaborative Filtering 

Alright, let's talk about how item-based collaborative filtering works. It's very similar to user-based collaborative filtering, but instead of users, we're looking at items.

So, let's go back to the example of movie recommendations. The first thing we would do is find every pair of movies that is watched by the same person. So, we go through and find every movie that was watched by identical people, and then we measure the similarity of all those people who viewed that movie to each other. So, by this means we can compute similarities between two different movies, based on the ratings of the people who watched both of those movies.

So, let's presume I have a movie pair, okay? Maybe Star Wars and The Empire Strikes Back. I find a list of everyone who watched both of those movies, then I compare their ratings to each other, and if they'r similar then I can say these two movies are similar, because they were rated similarly by people who watched both of them. That's the general idea here. That's one way to do it, there's more than one way to do it!

And then I can just sort everything by the movie, and then by the similarity strength of all the similar movies to it, and there's my results for people who liked also liked, or people who rated this highly also rated this highly and so on and so forth. And like I said, that's just one way of doing it. That's step one of item-based collaborative filtering-first I find relationships between movies based on the relationships of the people who watched every given pair of movies.

It'll make more sense when we go through the following example.

![title](Item_Based_Collaborative_Filtering_2.PNG)

For example, let's say that our nice young lady in the preceding image watched Star Wars and The Empire Strikes Back and liked both of them, so rated them both five stars or something. Now, along comes Mr. Edgy Mohawk Man who also watched Star Wars and The Empire Strikes Back and also liked both of them. So, at this point we can say there's a relationship, there is a similarity between Star Wars and The Empire Strikes Back based on these two users who liked both movies.

What we're going to do is look at each pair of movies. We have a pair of Star Wars and Empire Strikes Back, and then we look at all the users that watched both of them, which are these two guys, and if they both liked them, then we can say that they're similar to each other. Or, if they both disliked them we can also say they're similar to each other, right? So, we're just looking at the similarity score of these two users' behavior related to these two movies in this movie pair.

So, along comes Mr. Moustachy Lumberjack Hipster Man and he watches The Empire Strikes Back and he lives in some strange world where he watched The Empire Strikes Back, but had no idea that Star Wars the first movie existed.

![title](Item_Based_Collaborative_Filtering_3.PNG)

Well that's fine, we computed a relationship between The Empire Strikes Back and Star Wars based on the behavior of these two people, so we know that these two movies are similar to each other. So, given that Mr. Hipster Man liked The Empire Strikes Back, we can say with good confidence that he would also like Star Wars, and we can then recommend that back to him as his top movie recommendation.

## Collaborative Filtering Using Python

### Finding Movie Similarities

Let's apply the concept of item-based collaborative filtering. To start with, movie similarities-figure out what movies are similar to other movies. In particular, we'll try to figure out what movies are similar to Star Wars, based on user rating data, and we'll see what we get out of it.

We will be using some real movie rating data from the GroupLens project. GroupLens.org provides real movie ratings data, by real people who are using the MovieLens.org website to rate movies and get recommendations back for new movies that they want to watch.

The first thing we're going to do is import the u.data file as part of the MovieLens dataset, and that is a tabdelimited file that contains every rating in the dataset. We can specify a different separator than a comma. We're basically saying take the first three columns in the u.data file, and import it into a new DataFrame, with three columns: user_id, movie_id, and rating.

What we end up with here is a DataFrame that has a row for every user_id, which identifies some person, and then, for every movie they rated, we have the movie_id, which is some numerical shorthand for a given movie, so Star Wars might be movie 53 or something, and their rating, you know, 1 to 5 stars. So, we have here a database, a DataFrame, of every user and every movie they rated.

Now, we want to be able to work with movie titles, so we can interpret these results more intuitively, so we're going to use their human-readable names instead.

If you're using a truly massive dataset, you'd save that to the end because you want to be working with numbers, they're more compact, for as long as possible.

There's a separate data file with the MovieLens dataset called u.item, and it is pipe-delimited, and the first two columns that we import will be the movie_id and the title of that movie. So, now we have two DataFrames: r_cols has all the user ratings and m_cols has all the titles for every movie_id. We can then use the magical merge function in Pandas to mush it all together.

Let's add a ratings.head() command and then run those cells.

In [1]:
import pandas as pd
r_cols = ['user_id', 'movie_id', 'rating']
ratings = pd.read_csv('ml-100k/u.data', sep='\\t', names=r_cols, usecols=range(3), engine='python')

m_cols = ['movie_id', 'title']
movies = pd.read_csv('ml-100k/u.item', sep='|', names=m_cols, usecols=range(2), engine='python')

ratings = pd.merge(movies, ratings)

ratings.head()

movie_id             title  user_id  rating
0         1  Toy Story (1995)      308       4
1         1  Toy Story (1995)      287       5
2         1  Toy Story (1995)      148       4
3         1  Toy Story (1995)      280       4
4         1  Toy Story (1995)       66       3

We end up with a new DataFrame that contains the user_id and rating for each movie that a user rated, and we have both the movie_id and the title that we can read and see what it really is. So, the way to read this is user_id number 308 rated the Toy Story (1995) movie 4 stars, user_id number 287 rated the Toy Story (1995) movie 5 stars, and so on and so forth.

So, what we really want is to look at relationships between movies based on all the users that watched each pair of movies, so we need, at the end, a matrix of every movie, and every user, and all the ratings that every user gave to every movie. The pivot_table command in Pandas can do that for us. It can basically construct a new table from a given DataFrame, pretty much any way that you want it.

So, what we're saying with this code is-take our ratings DataFrame and create a new DataFrame called movieRatings and we want the index of it to be the user IDs, so we'll have a row for every user_id, and we're going to have every column be the movie title. So, we're going to have a column for every title that we encounter in that DataFrame, and each cell will contain the rating value, if it exists.

In [2]:
movieRatings = ratings.pivot_table(index=['user_id'], columns=['title'],values='rating')
movieRatings.head()

title    'Til There Was You (1997)  1-900 (1994)  101 Dalmatians (1996)  \
user_id                                                                   
0                              NaN           NaN                    NaN   
1                              NaN           NaN                    2.0   
2                              NaN           NaN                    NaN   
3                              NaN           NaN                    NaN   
4                              NaN           NaN                    NaN   

title    12 Angry Men (1957)  187 (1997)  2 Days in the Valley (1996)  \
user_id                                                                 
0                        NaN         NaN                          NaN   
1                        5.0         NaN                          NaN   
2                        NaN         NaN                          NaN   
3                        NaN         2.0                          NaN   
4                        NaN         NaN                          NaN   

title    20,000 Leagues Under the Sea (1954)  2001: A Space Odyssey (1968)  \
user_id                                                                      
0                                        NaN                           NaN   
1                                        3.0                           4.0   
2                                        NaN                           NaN   
3                                        NaN                           NaN   
4                                        NaN                           NaN   

title    3 Ninjas: High Noon At Mega Mountain (1998)  39 Steps, The (1935)  \
user_id                                                                      
0                                                NaN                   NaN   
1                                                NaN                   NaN   
2                                                1.0                   NaN   
3                                                NaN                   NaN   
4                                                NaN                   NaN   

title                   ...                  Yankee Zulu (1994)  \
user_id                 ...                                       
0                       ...                                 NaN   
1                       ...                                 NaN   
2                       ...                                 NaN   
3                       ...                                 NaN   
4                       ...                                 NaN   

title    Year of the Horse (1997)  You So Crazy (1994)  \
user_id                                                  
0                             NaN                  NaN   
1                             NaN                  NaN   
2                             NaN                  NaN   
3                             NaN                  NaN   
4                             NaN                  NaN   

title    Young Frankenstein (1974)  Young Guns (1988)  Young Guns II (1990)  \
user_id                                                                       
0                              NaN                NaN                   NaN   
1                              5.0                3.0                   NaN   
2                              NaN                NaN                   NaN   
3                              NaN                NaN                   NaN   
4                              NaN                NaN                   NaN   

title    Young Poisoner's Handbook, The (1995)  Zeus and Roxanne (1997)  \
user_id                                                                   
0                                          NaN                      NaN   
1                                          NaN                      NaN   
2                                          NaN                      NaN   
3                                          NaN                

So, the way to interpret this is, user_id number 1, for example, did not watch the movie 1-900 (1994), but user_id number 1 did watch 101 Dalmatians (1996) and rated it 2 stars. The user_id number 1 also watched 12 Angry Men (1957) and rated it 5 stars, but did not watch the movie 2 Days in the Valley (1996), for example, okay? So, what we end up with here is a sparse matrix basically, that contains every user, and every movie, and at every intersection where a user rated a movie there's a rating value.

So, you can see now, we can very easily extract vectors of every movie that our user watched, and we can also extract vectors of every user that rated a given movie, which is what we want. So, that's useful for both user-based and item-based collaborative filtering, right? If I wanted to find relationships between users, I could look at correlations between these user rows, but if I want to find correlations between movies, for item-based collaborative filtering, I can look at correlations between columns based on the user behavior.

let's go ahead and extract all the users who rated Star Wars (1977)

In [3]:
starWarsRatings = movieRatings['Star Wars (1977)']
starWarsRatings.head()

user_id
0    5.0
1    5.0
2    5.0
3    NaN
4    5.0
Name: Star Wars (1977), dtype: float64

And, we can see most people have, in fact, watched and rated Star Wars (1977) and everyone liked it, at least in this little sample that we took from the head of the DataFrame. So, we end up with a resulting set of user IDs and their ratings for Star Wars (1977). The user ID 3 did not rate Star Wars (1977) so we have a NaN value, indicating a missing value there, but that's okay. We want to make sure that we preserve those missing values so we can directly compare columns from different movies.

In order to compare columns with others we can use a function in python called corrwith. This will correlate a given column with every other column in the DataFrame, and compute the correlation scores and give that back to us. So, what we're doing here is using corrwith on the entire movieRatings DataFrame, that's that entire matrix of user movie ratings, correlating it with just the starWarsRatings column, and then dropping all of the missing results with dropna. So, that just leaves us with items that had a correlation, where there was more than one person that viewed it, and we create a new DataFrame based on those results and then display the top 10 results.

1. We're going to build the correlation score between Star Wars and every other movie.
2. Drop all the NaN values, so that we only have movie similarities that actually exist, where more than one person rated it.
3. And, we're going to construct a new DataFrame from the results and look at the top 10 results.

In [4]:
similarMovies = movieRatings.corrwith(starWarsRatings)
similarMovies = similarMovies.dropna()
df = pd.DataFrame(similarMovies)
df.head(10)

C:\Users\macharya\AppData\Local\Continuum\anaconda3\lib\site-packages\numpy\lib\function_base.py:2392: RuntimeWarning: Degrees of freedom <= 0 for slice
  c = cov(x, y, rowvar)
C:\Users\macharya\AppData\Local\Continuum\anaconda3\lib\site-packages\numpy\lib\function_base.py:2326: RuntimeWarning: divide by zero encountered in true_divide
  c *= np.true_divide(1, fact)


0
title                                        
'Til There Was You (1997)            0.872872
1-900 (1994)                        -0.645497
101 Dalmatians (1996)                0.211132
12 Angry Men (1957)                  0.184289
187 (1997)                           0.027398
2 Days in the Valley (1996)          0.066654
20,000 Leagues Under the Sea (1954)  0.289768
2001: A Space Odyssey (1968)         0.230884
39 Steps, The (1935)                 0.106453
8 1/2 (1963)                        -0.142977

We ended up with this result of correlation scores between each individual movie for Star Wars and we can see, for example, a surprisingly high correlation score with the movie 'Til There Was You (1997), a negative correlation with the movie 1-900 (1994), and a very weak correlation with 101 Dalmatians (1996). Now, all we should have to do is sort this by similarity score, and we should have the top movie similarities for Star Wars.

In [5]:
similarMovies.sort_values(ascending=False)

title
No Escape (1994)                                                                     1.000000
Man of the Year (1995)                                                               1.000000
Hollow Reed (1996)                                                                   1.000000
Commandments (1997)                                                                  1.000000
Cosi (1996)                                                                          1.000000
Stripes (1981)                                                                       1.000000
Golden Earrings (1947)                                                               1.000000
Mondo (1996)                                                                         1.000000
Line King: Al Hirschfeld, The (1996)                                                 1.000000
Outlaw, The (1943)                                                                   1.000000
Hurricane Streets (1998)                              

Okay, so Star Wars (1977) came out pretty close to top, because it is similar to itself, but what's all this other stuff? What the heck? We can see in the preceding output, some movies such as: Full Speed (1996), Man of the Year (1995), The Outlaw (1943). These are all, you know, fairly obscure movies, that most of them I've never even heard of, and yet they have perfect correlations with Star Wars. That's kinda weird! So, obviously we're doing something wrong here. What could it be?

So, just to remind you, we looked for movies that are similar to Star Wars using that technique, and we ended up with a bunch of weird recommendations at the top that had a perfect correlation. And, most of them were very obscure movies. So, what do you think might be going on there? Well, one thing that might make sense is, let's say we have a lot of people watch Star Wars and some other obscure film. We'd end up with a good correlation between these two movies because they're tied together by Star Wars, but at the end of the day, do we really want to base our recommendations on the behavior of one or two people that watch some obscure movie?

Probably Not. We need to have some sort of confidence level in our similarities by enforcing a minimum boundary of how many people watched a given movie. We can't make a judgment that a given movie is good just based on the behavior of one or two people. 

What we're going to do is try to identify the movies that weren't actually rated by many people and we'll just throw them out and see what we get. So, to do that we're going to take our original ratings DataFrame and we're going to say groupby('title'). And, this will basically construct a new DataFrame that aggregates together all the rows for a given title into one row. We can say that we want to aggregate specifically on the rating, and we want to show both the size, the number of ratings for each movie, and the mean average score, the mean rating for that movie.

In [6]:
import numpy as np
movieStats = ratings.groupby('title').agg({'rating': [np.size, np.mean]})
movieStats.head()

rating          
                            size      mean
title                                     
'Til There Was You (1997)      9  2.333333
1-900 (1994)                   5  2.600000
101 Dalmatians (1996)        109  2.908257
12 Angry Men (1957)          125  4.344000
187 (1997)                    41  3.024390

Let us take a cutoff to be 100. Let's go ahead and get rid of movies rated by fewer than 100 people.

In [7]:
popularMovies = movieStats['rating']['size'] >= 100
movieStats[popularMovies].sort_values([('rating', 'mean')], ascending=False)[:15]

rating          
                                         size      mean
title                                                  
Close Shave, A (1995)                     112  4.491071
Schindler's List (1993)                   298  4.466443
Wrong Trousers, The (1993)                118  4.466102
Casablanca (1942)                         243  4.456790
Shawshank Redemption, The (1994)          283  4.445230
Rear Window (1954)                        209  4.387560
Usual Suspects, The (1995)                267  4.385768
Star Wars (1977)                          584  4.359589
12 Angry Men (1957)                       125  4.344000
Citizen Kane (1941)                       198  4.292929
To Kill a Mockingbird (1962)              219  4.292237
One Flew Over the Cuckoo's Nest (1975)    264  4.291667
Silence of the Lambs, The (1991)          390  4.289744
North by Northwest (1959)                 179  4.284916
Godfather, The (1972)                     413  4.283293

We can just say popularMovies, a new DataFrame, is going to be constructed by looking at movieStats and we're going to only take rows where the rating size is greater than or equal to 100, and I'm then going to sort that by mean rating.

What we have here is a list of movies that were rated by more than 100 people, sorted by their average rating score, and this in itself is a recommender system. These are highly-rated popular movies. A Close Shave (1995), apparently, was a really good movie and a lot of people watched it and they really liked it.

Things look a little bit better now, so let's go ahead and basically make our new DataFrame of Star Wars recommendations, movies similar to Star Wars, where we only base it on movies that appear in this new DataFrame. So, we're going to use the join operation, to go ahead and join our original similarMovies DataFrame to this new DataFrame of only movies that have greater than 100 ratings.

In [8]:
df = movieStats[popularMovies].join(pd.DataFrame(similarMovies, columns=['similarity']))
df.head()

C:\Users\macharya\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\reshape\merge.py:544: UserWarning: merging between different levels can give an unintended result (2 levels on the left, 1 on the right)
  warnings.warn(msg, UserWarning)


(rating, size)  (rating, mean)  similarity
title                                                                   
101 Dalmatians (1996)                    109        2.908257    0.211132
12 Angry Men (1957)                      125        4.344000    0.184289
2001: A Space Odyssey (1968)             259        3.969112    0.230884
Absolute Power (1997)                    127        3.370079    0.085440
Abyss, The (1989)                        151        3.589404    0.203709

In the above code, we create a new DataFrame based on similarMovies where we extract the similarity column, join that with our movieStats DataFrame, which is our popularMovies DataFrame, and we look at the combined results.

Now we have, restricted only to movies that are rated by more than 100 people, the similarity score to Star Wars. So, now all we need to do is sort that using the following code.

In [9]:
df.sort_values(['similarity'], ascending=False)[:15]

(rating, size)  \
title                                                                
Star Wars (1977)                                               584   
Empire Strikes Back, The (1980)                                368   
Return of the Jedi (1983)                                      507   
Raiders of the Lost Ark (1981)                                 420   
Austin Powers: International Man of Mystery (1997)             130   
Sting, The (1973)                                              241   
Indiana Jones and the Last Crusade (1989)                      331   
Pinocchio (1940)                                               101   
Frighteners, The (1996)                                        115   
L.A. Confidential (1997)                                       297   
Wag the Dog (1997)                                             137   
Dumbo (1941)                                                   123   
Bridge on the River Kwai, The (1957)                           165   
Philadelphia Story, The (1940)                                 104   
Miracle on 34th Street (1994)                                  101   

                                                    (rating, mean)  similarity  
title                                                                           
Star Wars (1977)                                          4.359589    1.000000  
Empire Strikes Back, The (1980)                           4.206522    0.748353  
Return of the Jedi (1983)                                 4.007890    0.672556  
Raiders of the Lost Ark (1981)                            4.252381    0.536117  
Austin Powers: International Man of Mystery (1997)        3.246154    0.377433  
Sting, The (1973)                                         4.058091    0.367538  
Indiana Jones and the Last Crusade (1989)                 3.930514    0.350107  
Pinocchio (1940)                                          3.673267    0.347868  
Frighteners, The (1996)                                   3.234783    0.332729  
L.A. Confidential (1997)                                  4.161616    0.319065  
Wag the Dog (1997)                                        3.510949    0.318645  
Dumbo (1941)                                              3.495935    0.317656  
Bridge on the River Kwai, The (1957)                      4.175758    0.316580  
Philadelphia Story, The (1940)                            4.115385    0.314272  
Miracle on 34th Street (1994)                             3.722772    0.310921

This is starting to look a little bit better! So, Star Wars (1977) comes out on top because it's similar to itself, The Empire Strikes Back (1980) is number 2, Return of the Jedi (1983) is number 3, Raiders of the Lost Ark (1981), number 4. You know, it's still not perfect, but these make a lot more sense, right? So, you would expect the three Star Wars films from the original trilogy to be similar to each other, this data goes back to before the next three films, and Raiders of the Lost Ark (1981) is also a very similar movie to Star Wars in style, and i comes out as number 4. So, I'm starting to feel a little bit better about these results. There's still room for improvement, but hey! We got some results that make sense, whoo-hoo!

# Building a Recommender System

let's start off by importing the MovieLens dataset that we have. Again, we're using a subset of it that just contains 100,000 ratings for now. But, there are larger datasets you can get from GroupLens.org-up to millions of ratings; 

Just like earlier, we're going to import the u.data file that contains all the individual ratings for every user and what movie they rated, and then we're going to tie that together with the movie titles, so we don't have to just work with numerical movie IDs.

In [10]:
import pandas as pd
r_cols = ['user_id', 'movie_id', 'rating']
ratings = pd.read_csv('ml-100k/u.data', sep='\\t', names=r_cols, usecols=range(3), engine='python')

m_cols = ['movie_id', 'title']
movies = pd.read_csv('ml-100k/u.item', sep='|', names=m_cols, usecols=range(2), engine='python')

ratings = pd.merge(movies, ratings)

ratings.head()

movie_id             title  user_id  rating
0         1  Toy Story (1995)      308       4
1         1  Toy Story (1995)      287       5
2         1  Toy Story (1995)      148       4
3         1  Toy Story (1995)      280       4
4         1  Toy Story (1995)       66       3

And again, just like earlier, we use the wonderful pivot_table command in Pandas to construct a new DataFrame based on the information. Here, each row is the user_id, the columns are made up of all the unique movie titles in my dataset, and each cell contains a rating. 

In [43]:
userRatings = ratings.pivot_table(index=['user_id'], columns=['title'],values='rating')

userRatings.head()

title    'Til There Was You (1997)  1-900 (1994)  101 Dalmatians (1996)  \
user_id                                                                   
0                              NaN           NaN                    NaN   
1                              NaN           NaN                    2.0   
2                              NaN           NaN                    NaN   
3                              NaN           NaN                    NaN   
4                              NaN           NaN                    NaN   

title    12 Angry Men (1957)  187 (1997)  2 Days in the Valley (1996)  \
user_id                                                                 
0                        NaN         NaN                          NaN   
1                        5.0         NaN                          NaN   
2                        NaN         NaN                          NaN   
3                        NaN         2.0                          NaN   
4                        NaN         NaN                          NaN   

title    20,000 Leagues Under the Sea (1954)  2001: A Space Odyssey (1968)  \
user_id                                                                      
0                                        NaN                           NaN   
1                                        3.0                           4.0   
2                                        NaN                           NaN   
3                                        NaN                           NaN   
4                                        NaN                           NaN   

title    3 Ninjas: High Noon At Mega Mountain (1998)  39 Steps, The (1935)  \
user_id                                                                      
0                                                NaN                   NaN   
1                                                NaN                   NaN   
2                                                1.0                   NaN   
3                                                NaN                   NaN   
4                                                NaN                   NaN   

title                   ...                  Yankee Zulu (1994)  \
user_id                 ...                                       
0                       ...                                 NaN   
1                       ...                                 NaN   
2                       ...                                 NaN   
3                       ...                                 NaN   
4                       ...                                 NaN   

title    Year of the Horse (1997)  You So Crazy (1994)  \
user_id                                                  
0                             NaN                  NaN   
1                             NaN                  NaN   
2                             NaN                  NaN   
3                             NaN                  NaN   
4                             NaN                  NaN   

title    Young Frankenstein (1974)  Young Guns (1988)  Young Guns II (1990)  \
user_id                                                                       
0                              NaN                NaN                   NaN   
1                              5.0                3.0                   NaN   
2                              NaN                NaN                   NaN   
3                              NaN                NaN                   NaN   
4                              NaN                NaN                   NaN   

title    Young Poisoner's Handbook, The (1995)  Zeus and Roxanne (1997)  \
user_id                                                                   
0                                          NaN                      NaN   
1                                          NaN                      NaN   
2                                          NaN                      NaN   
3                                          NaN                

What we end up with is this incredibly useful matrix shown in the preceding output, that contains users for every row and movies for every column. And we have basically every user rating for every movie in this matrix. So, user_id number 1, for example, gave 101 Dalmatians (1996) a 2-star rating. And, again all these NaN values represent missing data. So, that just indicates, for example, user_id number 1 did not rate the movie 1-900 (1994).

Again, it's a very useful matrix to have. If we were doing user-based collaborative filtering, we could compute correlations between each individual user rating vector to find similar users. Since we're doing item-based collaborative filtering, we're more interested in relationships between the columns. So, for example, doing a correlation score between any two columns, which will give us a correlation score for a given movie pair.

The way we find the correlation in python is using the builtin function corr.

In [12]:
corrMatrix = userRatings.corr()
corrMatrix.head()

title                      'Til There Was You (1997)  1-900 (1994)  \
title                                                                
'Til There Was You (1997)                        1.0           NaN   
1-900 (1994)                                     NaN           1.0   
101 Dalmatians (1996)                           -1.0           NaN   
12 Angry Men (1957)                             -0.5           NaN   
187 (1997)                                      -0.5           NaN   

title                      101 Dalmatians (1996)  12 Angry Men (1957)  \
title                                                                   
'Til There Was You (1997)              -1.000000            -0.500000   
1-900 (1994)                                 NaN                  NaN   
101 Dalmatians (1996)                   1.000000            -0.049890   
12 Angry Men (1957)                    -0.049890             1.000000   
187 (1997)                              0.269191             0.666667   

title                      187 (1997)  2 Days in the Valley (1996)  \
title                                                                
'Til There Was You (1997)   -0.500000                     0.522233   
1-900 (1994)                      NaN                          NaN   
101 Dalmatians (1996)        0.269191                     0.048973   
12 Angry Men (1957)          0.666667                     0.256625   
187 (1997)                   1.000000                     0.596644   

title                      20,000 Leagues Under the Sea (1954)  \
title                                                            
'Til There Was You (1997)                                  NaN   
1-900 (1994)                                               NaN   
101 Dalmatians (1996)                                 0.266928   
12 Angry Men (1957)                                   0.274772   
187 (1997)                                                 NaN   

title                      2001: A Space Odyssey (1968)  \
title                                                     
'Til There Was You (1997)                     -0.426401   
1-900 (1994)                                  -0.981981   
101 Dalmatians (1996)                         -0.043407   
12 Angry Men (1957)                            0.178848   
187 (1997)                                    -0.554700   

title                      3 Ninjas: High Noon At Mega Mountain (1998)  \
title                                                                    
'Til There Was You (1997)                                          NaN   
1-900 (1994)                                                       NaN   
101 Dalmatians (1996)                                              NaN   
12 Angry Men (1957)                                                NaN   
187 (1997)                                                         NaN   

title                      39 Steps, The (1935)  \
title                                             
'Til There Was You (1997)                   NaN   
1-900 (1994)                                NaN   
101 Dalmatians (1996)                  0.111111   
12 Angry Men (1957)                    0.457176   
187 (1997)                             1.000000   

title                                     ...                  \
title                                     ...                   
'Til There Was You (1997)                 ...                   
1-900 (1994)                              ...                   
101 Dalmatians (1996)                     ...                   
12 Angry Men (1957)                       ...                   
187 (1997)                                ...                   

title                      Yankee Zulu (1994)  Year of the Horse (1997)  \
title                                                                     
'Til There Was You (1997)                 NaN                       NaN   
1-900 (1994)                              NaN                       NaN

We have here a new DataFrame where every movie is on the row, and in the column. So, we can look at the intersection of any two given movies and find their correlation score to each other based on this userRatings data that we had up here originally. For  example, the movie 101 Dalmatians (1996) is perfectly correlated with itself of course, because it has identical user rating vectors. But, if you look at 101 Dalmatians (1996) movie's relationship to the movie 12 Angry Men (1957), it's a much lower correlation score because those movies are rather dissimilar.

Now we have this wonderful matrix now that will give me the similarity score of any two movies to each other. Now just like earlier, we have to deal with spurious results. So, I don't want to be looking at relationships that are based on a small amount of behavior information.

It turns out that the Pandas corr function actually has a few parameters you can give it. One is the actual
correlation score method that you want to use, so I'm going to say use pearson correlation. You'll notice that it also has a min_periods parameter you can give it, and that basically says I only want you to consider correlation scores that are backed up by at least, in this example, 100 people that rated both movies. Running that will get rid of the spurious relationships that are based on just a handful of people.

In [13]:
corrMatrix = userRatings.corr(method='pearson', min_periods=100)
corrMatrix.head()

title                      'Til There Was You (1997)  1-900 (1994)  \
title                                                                
'Til There Was You (1997)                        NaN           NaN   
1-900 (1994)                                     NaN           NaN   
101 Dalmatians (1996)                            NaN           NaN   
12 Angry Men (1957)                              NaN           NaN   
187 (1997)                                       NaN           NaN   

title                      101 Dalmatians (1996)  12 Angry Men (1957)  \
title                                                                   
'Til There Was You (1997)                    NaN                  NaN   
1-900 (1994)                                 NaN                  NaN   
101 Dalmatians (1996)                        1.0                  NaN   
12 Angry Men (1957)                          NaN                  1.0   
187 (1997)                                   NaN                  NaN   

title                      187 (1997)  2 Days in the Valley (1996)  \
title                                                                
'Til There Was You (1997)         NaN                          NaN   
1-900 (1994)                      NaN                          NaN   
101 Dalmatians (1996)             NaN                          NaN   
12 Angry Men (1957)               NaN                          NaN   
187 (1997)                        NaN                          NaN   

title                      20,000 Leagues Under the Sea (1954)  \
title                                                            
'Til There Was You (1997)                                  NaN   
1-900 (1994)                                               NaN   
101 Dalmatians (1996)                                      NaN   
12 Angry Men (1957)                                        NaN   
187 (1997)                                                 NaN   

title                      2001: A Space Odyssey (1968)  \
title                                                     
'Til There Was You (1997)                           NaN   
1-900 (1994)                                        NaN   
101 Dalmatians (1996)                               NaN   
12 Angry Men (1957)                                 NaN   
187 (1997)                                          NaN   

title                      3 Ninjas: High Noon At Mega Mountain (1998)  \
title                                                                    
'Til There Was You (1997)                                          NaN   
1-900 (1994)                                                       NaN   
101 Dalmatians (1996)                                              NaN   
12 Angry Men (1957)                                                NaN   
187 (1997)                                                         NaN   

title                      39 Steps, The (1935)  \
title                                             
'Til There Was You (1997)                   NaN   
1-900 (1994)                                NaN   
101 Dalmatians (1996)                       NaN   
12 Angry Men (1957)                         NaN   
187 (1997)                                  NaN   

title                                     ...                  \
title                                     ...                   
'Til There Was You (1997)                 ...                   
1-900 (1994)                              ...                   
101 Dalmatians (1996)                     ...                   
12 Angry Men (1957)                       ...                   
187 (1997)                                ...                   

title                      Yankee Zulu (1994)  Year of the Horse (1997)  \
title                                                                     
'Til There Was You (1997)                 NaN                       NaN   
1-900 (1994)                              NaN                       NaN

It's a little bit different to what we did in the item similarities exercise where we just threw out any movie that was rated by less than 100 people. What we're doing here, is throwing out movie similarities where less than 100 people rated both of those movies.

In fact, even movies that are similar to themselves get thrown out, so for example, the movie 1-900 (1994) was, presumably, watched by fewer than 100 people so it just gets tossed entirely. The movie, 101 Dalmatians (1996) however, survives with a correlation score of 1, and there are actually no movies in this little sample of the dataset that are different from each other that had 100 people in common that watched both. But, there are enough movies that survive to get meaningful results.

## Understanding Movie Recommendation with an example.

What we want to do is recommend movies for people. The way we do that is we look at all the ratings for a given person, find movies similar to the stuff that they rated, and those are candidates for recommendations to that person.

Let's start by creating a fake person to create recommendations for. I've actually already added a fake user by hand, ID number 0, to the MovieLens dataset that we're processing. You can see that user with the following code

In [14]:
myRatings = userRatings.loc[0].dropna()
myRatings

title
Empire Strikes Back, The (1980)    5.0
Gone with the Wind (1939)          1.0
Star Wars (1977)                   5.0
Name: 0, dtype: float64

That kind of represents someone who loved Star Wars and The Empire Strikes Back, but hated the movie Gone with the Wind. So, this represents someone who really loves Star Wars, but does not like old style, romantic dramas, okay? So, we gave a rating of 5 star to The Empire Strikes Back (1980) and Star Wars (1977), and a rating of 1 star to Gone with the Wind (1939). So, we are going to try to find recommendations for this fictitious user.

Let's start by creating a series called simCandidates and I'm going to go through every movie that he rated. For i in range 0 through the number of ratings that we have in myRatings, we are going to add up similar movies to the ones that were rated. So, we are going to take that corrMatrix DataFrame, that one that has all of the movie similarities, and we are going to create a correlation matrix with myRatings, drop any missing values, and then we are going to scale that resulting correlation score by how well the movies were rated. So, the idea here is that we are going to go through all the similarities for The Empire Strikes Back, for example, and I will scale it all by 5, because I really liked The Empire Strikes Back. But, when I go through and get the similarities for Gone with the Wind, I'm only going to scale those by 1, because I did not like Gone with the Wind. So, this will give more strength to movies that are similar to movies that I liked, and less strength to movies that are similar to movies that I did not like,

In [30]:
simCandidates = pd.Series()
for i in range(0, len(myRatings.index)):
    print ("Adding sims for " + myRatings.index[i] + "...")
    # Retrieve similar movies to this one that I rated
    sims = corrMatrix[myRatings.index[i]].dropna()
    # Now scale its similarity by how well I rated this movie
    sims = sims.map(lambda x: x * myRatings[i])
    # Add the score to the list of similarity candidates
    simCandidates = simCandidates.append(sims)
#Glance at our results so far:
print ("sorting...")
simCandidates.sort_values(inplace = True, ascending = False)
print (simCandidates.head(10))

Adding sims for Empire Strikes Back, The (1980)...
Adding sims for Gone with the Wind (1939)...
Adding sims for Star Wars (1977)...
sorting...
Empire Strikes Back, The (1980)                       5.000000
Star Wars (1977)                                      5.000000
Empire Strikes Back, The (1980)                       3.741763
Star Wars (1977)                                      3.741763
Return of the Jedi (1983)                             3.606146
Return of the Jedi (1983)                             3.362779
Raiders of the Lost Ark (1981)                        2.693297
Raiders of the Lost Ark (1981)                        2.680586
Austin Powers: International Man of Mystery (1997)    1.887164
Sting, The (1973)                                     1.837692
dtype: float64


These results dont look so bad, The Empire Strikes Back (1980) and Star Wars (1977) come out on top, because those were rated high by out user. But, bubbling up to the top of the list is Return of the Jedi (1983), which we would expect and Raiders of the Lost Ark (1981).

Let's start to refine these results a little bit more. We're seeing that we're getting duplicate values back. If we have a movie that was similar to more than one movie that were rated, it will come back more than once in the results, so we want to combine those together. If I do in fact have the same movie, maybe that should get added up together into a combined, stronger recommendation score. Return of the Jedi, for example, was similar to both Star Wars and The Empire Strikes Back.

### Using Groupby Command to Combine Rows 

We're going to use the groupby command again to group together all of the rows that are for the same movie. Next, we will sum up their correlation score and look at the results.

In [31]:
simCandidates = simCandidates.groupby(simCandidates.index).sum()
simCandidates.sort_values(inplace = True, ascending = False)
simCandidates.head(10)

Empire Strikes Back, The (1980)              8.877450
Star Wars (1977)                             8.870971
Return of the Jedi (1983)                    7.178172
Raiders of the Lost Ark (1981)               5.519700
Indiana Jones and the Last Crusade (1989)    3.488028
Bridge on the River Kwai, The (1957)         3.366616
Back to the Future (1985)                    3.357941
Sting, The (1973)                            3.329843
Cinderella (1950)                            3.245412
Field of Dreams (1989)                       3.222311
dtype: float64

So Return of the Jedi (1983) comes out way on top, as it should, with a score of 7, Raiders of the Lost Ark (1981) a close second at 5, and then we start to get to Indiana Jones and the Last Crusade (1989), and some more movies, The Bridge on the River Kwai (1957), Back to the Future (1985),The Sting (1973).

The last thing we need to do is filter out the movies that were already rated, because it doesn't make sense to recommend movies you've already seen.

### Removing Entries with Drop Command

We can quickly drop any rows that happen to be in my original ratings series using the following code.

In [32]:
filteredSims = simCandidates.drop(myRatings.index)
filteredSims.head(10)

Return of the Jedi (1983)                    7.178172
Raiders of the Lost Ark (1981)               5.519700
Indiana Jones and the Last Crusade (1989)    3.488028
Bridge on the River Kwai, The (1957)         3.366616
Back to the Future (1985)                    3.357941
Sting, The (1973)                            3.329843
Cinderella (1950)                            3.245412
Field of Dreams (1989)                       3.222311
Wizard of Oz, The (1939)                     3.200268
Dumbo (1941)                                 2.981645
dtype: float64

And there we have it! Return of the Jedi (1983), Raiders of the Lost Ark (1981), Indiana Jones and the Last Crusade (1989), all the top results for my fictitious user, and they all make sense. I'm seeing a few family-friendly films, you know, Cinderella (1950), The Wizard of Oz (1939), Dumbo (1941), creeping in, probably based on the presence of Gone with the Wind in there, even though it was weighted downward it's still in there, and still being counted. And, there we have our results.

In [ ]:
def movierating(period,subject,scaling):
    corrMatrix = userRatings.corr(method='pearson', min_periods = period)
    myRatings = userRatings.loc[subject].dropna()
    if scaling = 1:
        simCandidates = pd.Series()
        for i in range(0, len(myRatings.index)):
            print ("Adding sims for " + myRatings.index[i] + "...")
            # Retrieve similar movies to this one that I rated
            sims = corrMatrix[myRatings.index[i]].dropna()
            # Now scale its similarity by how well I rated this movie # We are scaling this by a Square function which weights
            # larger rating more than lower rating
            sims = sims.map(lambda x: x * np.log10(myRatings[i]))
            # Add the score to the list of similarity candidates
            simCandidates = simCandidates.append(sims)
        #Glance at our results so far:
        print ("sorting...")
        simCandidates.sort_values(inplace = True, ascending = False)
        print (simCandidates.head(10))
    elif scaling = 2:
        # Here using square function for scaling
        simCandidates = pd.Series()
        for i in range(0, len(myRatings.index)):
            print ("Adding sims for " + myRatings.index[i] + "...")
            # Retrieve similar movies to this one that I rated
            sims = corrMatrix[myRatings.index[i]].dropna()
            # Now scale its similarity by how well I rated this movie # We are scaling this by a Square function which weights
            # larger rating more than lower rating
            sims = sims.map(lambda x: x * myRatings[i] * myRatings[i])
            # Add the score to the list of similarity candidates
            simCandidates = simCandidates.append(sims)
        #Glance at our results so far:
        print ("sorting...")
        simCandidates.sort_values(inplace = True, ascending = False)
        print (simCandidates.head(10))
    else :
        simCandidates = pd.Series()
        for i in range(0, len(myRatings.index)):
            print ("Adding sims for " + myRatings.index[i] + "...")
            # Retrieve similar movies to this one that I rated
            sims = corrMatrix[myRatings.index[i]].dropna()
            # Now scale its similarity by how well I rated this movie # We are scaling this by a Square function which weights
            # larger rating more than lower rating
            sims = sims.map(lambda x: x * myRatings[i] * myRatings[i] / 25.0)
            # Add the score to the list of similarity candidates
            simCandidates = simCandidates.append(sims)
        #Glance at our results so far:
        print ("sorting...")
        simCandidates.sort_values(inplace = True, ascending = False)
        print (simCandidates.head(10))
    simCandidates = simCandidates.groupby(simCandidates.index).sum()
    simCandidates.sort_values(inplace = True, ascending = False)
    filteredSims = simCandidates.drop(myRatings.index)

